hudi-common/src/main/java/org/apache/hudi/io/storage/HFileReaderFactory.java

```diff
@@ -55,9 +58,34 @@ public HFileReaderFactory(HoodieStorage storage,
   public HFileReader createHFileReader() throws IOException {
     final long fileSize = determineFileSize();
     final SeekableDataInputStream inputStream = createInputStream(fileSize);
+    
+    if (shouldEnableBlockCaching()) {
+      HFileReaderConfig config = createHFileReaderConfig();
+      String filePath = getFilePath();
+      return new CachingHFileReaderImpl(inputStream, fileSize, filePath, config);
+    }
+    
     return new HFileReaderImpl(inputStream, fileSize);
   }
 
+  private boolean shouldEnableBlockCaching() {
+    return metadataConfig.getHFileBlockCacheEnabled();
+  }
+
+  private HFileReaderConfig createHFileReaderConfig() {
+    int blockCacheSize = metadataConfig.getHFileBlockCacheSize();
+    int cacheTtlMinutes = metadataConfig.getHFileBlockCacheTTLMinutes();
+    return new HFileReaderConfig(blockCacheSize, cacheTtlMinutes);
+  }
+
+  private String getFilePath() {
+    if (fileSource.isLeft()) {
+      return fileSource.asLeft().toString();
+    }
+    // For byte array content, use a hash-based identifier
+    return "bytes:" + Arrays.hashCode(fileSource.asRight());
```

Comment:
What is the overhead of computing the hash code of a large byte array from a large HFile block?

Reply:
Computation should be linear to length of byte array with we use hashCode. 

Will change to:

```
Hashing.murmur3_32_fixed().hashBytes(byteArray).asInt();
```

Reply:
Changed this to:
```
int murmurHash = MurmurHash.getInstance().hash(fileSource.asRight());
return String.valueOf(murmurHash);
```

From what i understand, It's better than `Arrays.hashCode` it does chunk processing instead of a byte by byte processing although both of them have O(N) complexity.

Reply:
Let's track this in a separate PR too.